src/domain/session/room/lightbox-navigation.js

```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2022 Bruno Windels <bruno@windels.cloud>
+Copyright 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import {LightboxViewModel} from "./LightboxViewModel.js";
+
+// Store the `LightboxViewModel` under a symbol so no one else can tamper with
+// it. This acts like a private field on the class since no one else has the
+// symbol to look it up.
+let lightboxViewModelSymbol = Symbol('lightboxViewModel');
+
+/**
+ * Destroys and creates a new the `LightboxViewModel` depending if
+ * `lightboxChildOptions.eventEntry` or `lightboxChildOptions.eventId` are
+ * provided.
+ */
+function updateLightboxViewModel(vm, fieldName, lightboxChildOptions) {
```

Comment:
To make it make sense it my head, here would be the usage for `lightbox` (adapted to using object parameters since there is so many flying around here):

```js
this.bindChildToNavigationSegment({
  segmentName: 'lightbox',
  setter: (newValue) => {
    this._lightboxViewModel = newValue;
  },
  getter: () = {
    return this._lightboxViewModel;
  },
  vmMapper: (eventId) = {
    return new LightboxViewModel(this.childOptions({eventId, room}));
  },
  publicPropName: 'lightboxViewModel'
});
```


`bindChildToNavigationSegment` does make a lot of sense to capture this abstraction because it's so common.

The usage seems like too much boilerplate for what a SDK consumer cares about though.


---

I think we could boil it down to:

```js
this.bindChildToNavigationSegment({
  segmentName: 'lightbox',
  vmMapper: (eventId) = {
    return new LightboxViewModel(this.childOptions({eventId, room}));
  },
  publicPropName: 'lightboxViewModel'
});
```

```js
class ViewModel {
	bindChildToNavigationSegment(segmentName, vmMapper, publicPropName) {
        // Store the `ViewModel` under a symbol so no one else can tamper with
        // it. This acts like a private field on the class since no one else has the
        // symbol to look it up.
	    let viewModelSymbol = Symbol(publicPropName);

	    // On the given `vm`, create a getter at `publicPropName` that the
        // `ViewModel` is exposed at for usage in the view.
        Object.defineProperty(vm, publicPropName, {
            get: function() {
                return vm[viewModelSymbol];
            }
        });
        
		const update = value => {
			if (vm[lightboxViewModelSymbol]) {
	            vm[lightboxViewModelSymbol] = this.disposeTracked(getter());
	        }
	        if (value) {
	            vm[lightboxViewModelSymbol] = this.track(vmMapper(value));
	        }
	        this.emitChange(publicPropName);
		}

		const observable = this.navigation.observe(segmentName);
        this.track(observable.subscribe(update));
        update(observable.get());
	}
}
```

If some use case needs `getter`/`setter` we could have those as extra options and store it for them by default as above ^

Reply:
@bwindels Does the boiled down version look reasonable?